overlord/state/taskrunner.go

```diff
@@ -84,6 +104,22 @@ func (r *TaskRunner) AddHandler(kind string, do, undo HandlerFunc) {
 	r.handlers[kind] = handlerPair{do, undo}
 }
 
+func (r *TaskRunner) getHandlerPair(kind string) handlerPair {
+	handlers, ok := r.handlers[kind]
+	if !ok {
+		// Should we capture unknown tasks?
+		if len(r.knownKinds) > 0 && !r.knownKinds[kind] {
+			handlers = handlerPair{
+				do: func(task *Task, tomb *tomb.Tomb) error {
+					return fmt.Errorf("no handler for task %q", kind)
```

Comment:
This is not what we agreed to do, I think. If I remember correctly we said that given the cases we know about for now, we'd like to skip over the cases we don't know about. This perhaps also shows an issue in the current strategy: it over-specializes the logic to a particular case. It would be nicer to introduce an extension mechanism that allows the call site to use it in whatever way it pleases.

How about something like this: in addition to AddHandler, we introduce:
```
func (r *TaskRunner) AddOptionalHandler(match func(t *Task) bool, do, undo HandlerFunc) {
        t.optional = append(t.optional, optionalHandler{match, handlerPair{do, undo}})
}
```
We keep these as
```
type optionalHandler struct {
        match func(t *Task) bool
        handlerPair
}

type TaskRunner {
        ...
        handlers map[string]handlerPair
        optional []optionalHandler
        ...
}
```

This means we can reestructure the logic so that these handlers are also normal handlers, positive instead of negative as it looks like at the moment. The way we use it externally is what makes it possible to tackle unknown tasks, but we don't care here.

Reply:
I like this approach. I had different idea about future enhancements to this functionality - not modular and not extensible externally; I agree what you suggested is cleaner and nicer, thanks for the idea. I've implemented these changes.